ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py

```diff
@@ -0,0 +1,167 @@
+    return authenticated_client
+
+def get_object(object_storage_client, namespace, bucket, file):
+    get_resp = object_storage_client.get_object(namespace, bucket, file)
+    return get_resp.data.text
+
+def parse_and_process_data_preprocessing_config(object_storage_client, spark, get_resp, output_path):
+    dfs = dict()
+    try:
+        contents = json.loads(get_resp)
+        '''
+        Deal with input sources
+        '''
+        input_sources = contents["inputSources"]
+
+        # build dictonary: df name: df
+        for source in input_sources:
+            if source["type"] == "object-storage":
+                raw_data = get_object(object_storage_client, source["namespace"], source["bucket"], source["objectName"])
+                data = StringIO(raw_data)
+                pd_df = pd.read_csv(data, sep=",")
+                df = spark.createDataFrame(pd_df)
+                df = df.select([F.col(x).alias(x.lower()) for x in df.columns])
+                dfs[source["dataframeName"]] = df
+            elif source["type"] == "database":
+                properties = {
+                    "adbId": source["adbId"], 
+                    "dbtable": source["tableName"],
+                    "connectionId": source["connectionId"],
+                    "user": source["user"],
+                    "password": source["password"]}
+                df = spark.read \
+                    .format("oracle") \
+                    .options(**properties) \
+                    .load()
+                df = df.select([F.col(x).alias(x.lower()) for x in df.columns])
+                dfs[source["dataframeName"]] = df
+
+        '''
+        Check the phase
+        train: hold on until the end of processing steps, saving reserved global variables to metadata
+        inference: load the metadata into global variables
+        '''
+        metadata = dict()
+        phaseInfo = contents["phaseInfo"]
+        phase = phaseInfo["phase"]
+        if phase == "inference":
+            metadata_dependent_raw = get_object(object_storage_client, phaseInfo["connector"]["namespace"], phaseInfo["connector"]["bucket"], phaseInfo["connector"]["objectName"])
+            metadata_dependent = json.loads(metadata_dependent_raw)
+            for global_variable in RESERVED_METADATA:
+                metadata[global_variable] = metadata_dependent[global_variable]
+        elif phase == "train":
+            pass
+        else:
+            raise Exception("phaseInfo is not correct") 
+
+
+        # conducting processing steps
+        processing_steps = contents["processingSteps"]
+        for step_config in processing_steps:
+            if step_config["stepType"] == SINGLE_DATAFRAME_PROCESSING:
+                for step in step_config["configurations"]["steps"]:
+                    func_name = step["stepName"]
+                    # one_hot_encoding is speicifc because it is data dependent transformation
+                    if func_name == "one_hot_encoding":
```

Comment:
let's keep a list of data dependent transformations, and add both one-hot encoding and pivoting to the list

Reply:
From coding wise, it's meaningless to put them in a list, since different data dependent transforms need to read different metadata. I think it's better to add data dependent information into document